lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomain returns the protection domain behind domainID.
+func (p *powerFlexClient) getProtectionDomain(domainID string) (*powerFlexProtectionDomain, error) {
+	var actualResponse powerFlexProtectionDomain
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain: %q: %w", domainID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomainSdsRelations returns the protection domains SDS relations.
+func (p *powerFlexClient) getProtectionDomainSdsRelations(domainID string) ([]powerFlexProtectionDomainSdsRelation, error) {
+	var actualResponse []powerFlexProtectionDomainSdsRelation
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s/relationships/Sds", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain SDS relations: %q: %w", domainID, err)
+	}
+
+	return actualResponse, nil
+}
+
+// getVolumeID returns the volume ID for the given name.
+func (p *powerFlexClient) getVolumeID(name string) (string, error) {
+	body, err := p.createBodyReader(map[string]any{
+		"name": name,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	var actualResponse string
+	err = p.requestAuthenticated(http.MethodPost, "/api/types/Volume/instances/action/queryIdByKey", body, &actualResponse)
+	if err != nil {
+		powerFlexError, ok := err.(*powerFlexError)
+		if ok {
+			// API returns 500 if the volume does not exist.
+			// To not confuse it with other 500 that might occur check the error code too.
+			if powerFlexError.HTTPStatusCode() == 500 && powerFlexError.ErrorCode() == 79 {
+				return "", api.StatusErrorf(http.StatusNotFound, "PowerFlex volume not found")
+			}
+		}
+
+		return "", fmt.Errorf("Failed to get volume ID: %q: %w", name, err)
+	}
+
+	return actualResponse, nil
+}
+
+// getVolume returns the volume behind volumeID.
+func (p *powerFlexClient) getVolume(volumeID string) (*powerFlexVolume, error) {
+	var actualResponse powerFlexVolume
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/Volume::%s", volumeID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get volume: %q: %w", volumeID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// createVolume creates a new volume.
+// The returned string represents the ID of the volume.
+func (p *powerFlexClient) createVolume(name string, size string, volumeType string, poolID string) (string, error) {
+	switch volumeType {
+	case "thin":
+		volumeType = "ThinProvisioned"
+	case "thick":
+		volumeType = "ThickProvisioned"
+	default:
+		return "", fmt.Errorf("Invalid volume type: %q", volumeType)
+	}
+
+	body, err := p.createBodyReader(map[string]any{
+		"name":           name,
+		"volumeSizeInGb": size,
```

Comment:
Is "Gb" here "GB" or "GiB"? We're going to be need to be careful about this.

Does it accept a value in bytes btw?

Reply:
Even though the request body says `Gb`, the value is interpreted as GiB. Same applies for the PowerFlex UI were the volume size is printed with `GB` but the actual value has unit GiB internally. I validated it by checking the actual volume size on the system after adding the disks.

There are no other units than GiB supported by the API.

Reply:
A comment to that effect would be good here then to save that knowledge out of your head and into the code :)

Reply:
Comment added.